packages/migrations/src/utils/edtr-io.ts

```diff
@@ -0,0 +1,69 @@
+/**
+ * This file is part of Serlo.org.
+ *
+ * Copyright (c) 2013-2021 Serlo Education e.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @copyright Copyright (c) 2013-2021 Serlo Education e.V.
+ * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://github.com/serlo-org/serlo.org for the canonical source repository
+ */
+import * as R from 'ramda'
+
+export function updatePlugins(
+  transformations: {
+    [key in string]?: (args: {
+      state: unknown
+      transformState: Transformation
+    }) => unknown
+  }
+): Transformation {
+  function transformState(value: unknown): unknown {
+    if (isPlugin(value)) {
+      const { plugin, state } = value
+      const transformFunc = transformations[plugin]
+
+      if (transformFunc !== undefined) {
+        return { ...value, state: transformFunc({ state, transformState }) }
+      }
+    }
+
+    if (Array.isArray(value)) {
+      return value.map(transformState)
+    }
+
+    if (typeof value === 'object' && value !== null) {
+      return R.mapObjIndexed(transformState, value)
+    }
```

Comment:
That looks a little bit scary. Why don't you do the thing we did for the other migrations, i.e. actually going over the plugins where we know that they should have children? This will fail if we use the key `plugin` in some plugin's state or substate.

Reply:
Also: does this even work if `state = undefined`?

Reply:
The beauty of it is that the code works regardless of the plugins we currently. Besides no state should have a `plugin` + `state` property...

Reply:
The problem with it is that it works on implicit assumptions. Especially for third-party code, the assumption might not hold. For example: are you sure that Slate's state doesn't include "plugin" + "state"?

Since we want to go over the migration process anyways and we move Edtr.io to the frontend soon-ish, I'm fine with merging this. We should think about another long-term solution, though.